`sqlx-postgres` Fails to Build on Latest 1.91.0

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
// sqlx library
```

I expected compilation to succeed.
Instead, this happened: hangs for a while (possibly due to recursion) and then errors out with 
```error[E0275]: overflow evaluating the requirement `(&mut PgConnection, Oid): std::marker::Send` ```.

<details>

```
error[E0275]: overflow evaluating the requirement `(&mut PgConnection, Oid): std::marker::Send`
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-core-0.8.6\src\ext\async_stream.rs:28:56
     |
  28 |         Fut: 'a + Future<Output = Result<(), Error>> + Send,
     |                                                        ^^^^
     |
     = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`sqlx_postgres`)
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:319:36
     |
 319 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 320 | |         let base_type = self.maybe_fetch_type_info_by_oid(base_type, true).await?;
 321 | |
 322 | |         Ok(PgTypeInfo(PgType::Custom(Arc::new(PgCustomType {
...    |
 326 | |         }))))
 327 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:193:82
     |
 193 |       async fn fetch_type_by_oid(&mut self, oid: Oid) -> Result<PgTypeInfo, Error> {
     |  __________________________________________________________________________________^
 194 | |         let (name, typ_type, category, relation_id, element, base_type): (
 195 | |             String,
 196 | |             i8,
...    |
 257 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33
     |
 172 |             let info = Box::pin(async { self.fetch_type_by_oid(oid).await }).await?;
     |                                 ^^^^^
     = note: required for `Unique<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>` to implement `std::marker::Send`
note: required because it appears within the type `Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:231:12
     |
 231 | pub struct Box<
     |            ^^^
note: required because it appears within the type `Pin<Box<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:172:33: 172:38}>>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\pin.rs:1094:12
     |
1094 | pub struct Pin<Ptr> {
     |            ^^^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:157:36
     |
 157 |       ) -> Result<PgTypeInfo, Error> {
     |  ____________________________________^
 158 | |         // first we check if this is a built-in type
 159 | |         // in the average application, the vast majority of checks should flow through this
 160 | |         if let Some(info) = PgTypeInfo::try_from_oid(oid) {
...    |
 191 | |     }
     | |_____^
note: required because it's used within this `async` fn body
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\describe.rs:104:63
     |
 104 |       ) -> Result<(Vec<PgColumn>, HashMap<UStr, usize>), Error> {
     |  _______________________________________________________________^
 105 | |         let mut columns = Vec::new();
 106 | |         let mut column_names = HashMap::new();
...    |
 137 | |         Ok((columns, column_names))
 138 | |     }
     | |_____^
note: required because it's used within this `async` block
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-postgres-0.8.6\src\connection\executor.rs:294:12
     |
 294 |           Ok(try_stream! {
     |  ____________^
 295 | |             loop {
 296 | |                 let message = self.inner.stream.recv().await?;
...    |
 375 | |             Ok(())
 376 | |         })
     | |_________^
note: required because it appears within the type `ManuallyDrop<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-core-0.8.6\src\ext\async_stream.rs:124:110: 124:120}>`
    --> C:\Users\htgaz\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\mem\manually_drop.rs:157:12
     |
 157 | pub struct ManuallyDrop<T: ?Sized> {
     |            ^^^^^^^^^^^^
note: required because it appears within the type `Instrumented<{async block@C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\sqlx-core-0.8.6\src\ext\async_stream.rs:124:110: 124:120}>`
    --> C:\Users\htgaz\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\tracing-0.1.41\src\instrument.rs:266:16
     |
 266 |     pub struct Instrumented<T> {
     |                ^^^^^^^^^^^^
     = note: this error originates in the macro `try_stream` (in Nightly builds, run with -Z macro-backtrace for more info)
```

</details>

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.91.0-nightly (1ebbd87a6 2025-08-11)
binary: rustc
commit-hash: 1ebbd87a62ce96a72b22da61b7c2c43893534842
commit-date: 2025-08-11
host: x86_64-pc-windows-msvc
release: 1.91.0-nightly
LLVM version: 21.1.0
```

--- a/compiler/rustc_middle/src/ty/sty.rs
+++ b/compiler/rustc_middle/src/ty/sty.rs
@@ -821,10 +821,13 @@ impl<'tcx> Ty<'tcx> {
     #[inline]
     pub fn new_coroutine_witness(
         tcx: TyCtxt<'tcx>,
-        id: DefId,
+        def_id: DefId,
         args: GenericArgsRef<'tcx>,
     ) -> Ty<'tcx> {
-        Ty::new(tcx, CoroutineWitness(id, args))
+        if cfg!(debug_assertions) {
+            tcx.debug_assert_args_compatible(tcx.typeck_root_def_id(def_id), args);
+        }
+        Ty::new(tcx, CoroutineWitness(def_id, args))
     }
 
     // misc

--- a/compiler/rustc_middle/src/ty/sty.rs
+++ b/compiler/rustc_middle/src/ty/sty.rs
@@ -830,6 +830,31 @@ impl<'tcx> Ty<'tcx> {
         Ty::new(tcx, CoroutineWitness(def_id, args))
     }
 
+    pub fn new_coroutine_witness_for_coroutine(
+        tcx: TyCtxt<'tcx>,
+        def_id: DefId,
+        coroutine_args: GenericArgsRef<'tcx>,
+    ) -> Ty<'tcx> {
+        tcx.debug_assert_args_compatible(def_id, coroutine_args);
+        // HACK: Coroutine witness types are lifetime erased, so they
+        // never reference any lifetime args from the coroutine. We erase
+        // the regions here since we may get into situations where a
+        // coroutine is recursively contained within itself, leading to
+        // witness types that differ by region args. This means that
+        // cycle detection in fulfillment will not kick in, which leads
+        // to unnecessary overflows in async code. See the issue:
+        // <https://github.com/rust-lang/rust/issues/145151>.
+        let args =
+            ty::GenericArgs::for_item(tcx, tcx.typeck_root_def_id(def_id), |def, _| {
+                match def.kind {
+                    ty::GenericParamDefKind::Lifetime => tcx.lifetimes.re_erased.into(),
+                    ty::GenericParamDefKind::Type { .. }
+                    | ty::GenericParamDefKind::Const { .. } => coroutine_args[def.index as usize],
+                }
+            });
+        Ty::new_coroutine_witness(tcx, def_id, args)
+    }
+
     // misc
 
     #[inline]
@@ -986,6 +1011,14 @@ impl<'tcx> rustc_type_ir::inherent::Ty<TyCtxt<'tcx>> for Ty<'tcx> {
         args: ty::GenericArgsRef<'tcx>,
     ) -> Self {
         Ty::new_coroutine_witness(interner, def_id, args)
+    }
+
+    fn new_coroutine_witness_for_coroutine(
+        interner: TyCtxt<'tcx>,
+        def_id: DefId,
+        coroutine_args: ty::GenericArgsRef<'tcx>,
+    ) -> Self {
+        Ty::new_coroutine_witness_for_coroutine(interner, def_id, coroutine_args)
     }
 
     fn new_ptr(interner: TyCtxt<'tcx>, ty: Self, mutbl: hir::Mutability) -> Self {

--- a/compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs
+++ b/compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs
@@ -75,17 +75,10 @@ where
             Ok(ty::Binder::dummy(vec![args.as_coroutine_closure().tupled_upvars_ty()]))
         }
 
-        ty::Coroutine(def_id, args) => {
-            let coroutine_args = args.as_coroutine();
-            Ok(ty::Binder::dummy(vec![
-                coroutine_args.tupled_upvars_ty(),
-                Ty::new_coroutine_witness(
-                    ecx.cx(),
-                    def_id,
-                    ecx.cx().mk_args(coroutine_args.parent_args().as_slice()),
-                ),
-            ]))
-        }
+        ty::Coroutine(def_id, args) => Ok(ty::Binder::dummy(vec![
+            args.as_coroutine().tupled_upvars_ty(),
+            Ty::new_coroutine_witness_for_coroutine(ecx.cx(), def_id, args),
+        ])),
 
         ty::CoroutineWitness(def_id, args) => Ok(ecx
             .cx()
@@ -251,14 +244,9 @@ where
             Movability::Static => Err(NoSolution),
             Movability::Movable => {
                 if ecx.cx().features().coroutine_clone() {
-                    let coroutine = args.as_coroutine();
                     Ok(ty::Binder::dummy(vec![
-                        coroutine.tupled_upvars_ty(),
-                        Ty::new_coroutine_witness(
-                            ecx.cx(),
-                            def_id,
-                            ecx.cx().mk_args(coroutine.parent_args().as_slice()),
-                        ),
+                        args.as_coroutine().tupled_upvars_ty(),
+                        Ty::new_coroutine_witness_for_coroutine(ecx.cx(), def_id, args),
                     ]))
                 } else {
                     Err(NoSolution)

--- a/compiler/rustc_trait_selection/src/traits/select/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/select/mod.rs
@@ -4,9 +4,9 @@
 
 use std::assert_matches::assert_matches;
 use std::cell::{Cell, RefCell};
+use std::cmp;
 use std::fmt::{self, Display};
 use std::ops::ControlFlow;
-use std::{cmp, iter};
 
 use hir::def::DefKind;
 use rustc_data_structures::fx::{FxIndexMap, FxIndexSet};
@@ -2185,32 +2185,23 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
                 ty::Binder::dummy(vec![ty])
             }
 
-            ty::Coroutine(coroutine_def_id, args) => {
-                match self.tcx().coroutine_movability(coroutine_def_id) {
-                    hir::Movability::Static => {
-                        unreachable!("tried to assemble `Sized` for static coroutine")
-                    }
-                    hir::Movability::Movable => {
-                        if self.tcx().features().coroutine_clone() {
-                            ty::Binder::dummy(
-                                args.as_coroutine()
-                                    .upvar_tys()
-                                    .iter()
-                                    .chain([Ty::new_coroutine_witness(
-                                        self.tcx(),
-                                        coroutine_def_id,
-                                        self.tcx().mk_args(args.as_coroutine().parent_args()),
-                                    )])
-                                    .collect::<Vec<_>>(),
-                            )
-                        } else {
-                            unreachable!(
-                                "tried to assemble `Sized` for coroutine without enabled feature"
-                            )
-                        }
+            ty::Coroutine(def_id, args) => match self.tcx().coroutine_movability(def_id) {
+                hir::Movability::Static => {
+                    unreachable!("tried to assemble `Clone` for static coroutine")
+                }
+                hir::Movability::Movable => {
+                    if self.tcx().features().coroutine_clone() {
+                        ty::Binder::dummy(vec![
+                            args.as_coroutine().tupled_upvars_ty(),
+                            Ty::new_coroutine_witness_for_coroutine(self.tcx(), def_id, args),
+                        ])
+                    } else {
+                        unreachable!(
+                            "tried to assemble `Clone` for coroutine without enabled feature"
+                        )
                     }
                 }
-            }
+            },
 
             ty::CoroutineWitness(def_id, args) => self
                 .infcx
@@ -2334,25 +2325,9 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
             ty::Coroutine(def_id, args) => {
                 let ty = self.infcx.shallow_resolve(args.as_coroutine().tupled_upvars_ty());
                 let tcx = self.tcx();
-                let witness = Ty::new_coroutine_witness(
-                    tcx,
-                    def_id,
-                    ty::GenericArgs::for_item(tcx, def_id, |def, _| match def.kind {
-                        // HACK: Coroutine witnesse types are lifetime erased, so they
-                        // never reference any lifetime args from the coroutine. We erase
-                        // the regions here since we may get into situations where a
-                        // coroutine is recursively contained within itself, leading to
-                        // witness types that differ by region args. This means that
-                        // cycle detection in fulfillment will not kick in, which leads
-                        // to unnecessary overflows in async code. See the issue:
-                        // <https://github.com/rust-lang/rust/issues/145151>.
-                        ty::GenericParamDefKind::Lifetime => tcx.lifetimes.re_erased.into(),
-                        ty::GenericParamDefKind::Type { .. }
-                        | ty::GenericParamDefKind::Const { .. } => args[def.index as usize],
-                    }),
-                );
+                let witness = Ty::new_coroutine_witness_for_coroutine(tcx, def_id, args);
                 ty::Binder::dummy(AutoImplConstituents {
-                    types: [ty].into_iter().chain(iter::once(witness)).collect(),
+                    types: vec![ty, witness],
                     assumptions: vec![],
                 })
             }

--- a/compiler/rustc_type_ir/src/inherent.rs
+++ b/compiler/rustc_type_ir/src/inherent.rs
@@ -91,6 +91,12 @@ pub trait Ty<I: Interner<Ty = Self>>:
 
     fn new_coroutine_witness(interner: I, def_id: I::DefId, args: I::GenericArgs) -> Self;
 
+    fn new_coroutine_witness_for_coroutine(
+        interner: I,
+        def_id: I::DefId,
+        coroutine_args: I::GenericArgs,
+    ) -> Self;
+
     fn new_ptr(interner: I, ty: Self, mutbl: Mutability) -> Self;
 
     fn new_ref(interner: I, region: I::Region, ty: Self, mutbl: Mutability) -> Self;

--- a/tests/ui/async-await/recursive-async-auto-trait-overflow-only-parent-args.rs
+++ b/tests/ui/async-await/recursive-async-auto-trait-overflow-only-parent-args.rs
@@ -1,0 +1,17 @@
+// Regression test for #145288. This is the same issue as #145151
+// which we fixed in #145194. However in that PR we accidentally created
+// a `CoroutineWitness` which referenced all generic arguments of the
+// coroutine, including upvars and the signature.
+
+//@ edition: 2024
+//@ check-pass
+
+async fn process<'a>(x: &'a u32) {
+    Box::pin(process(x)).await;
+}
+
+fn require_send(_: impl Send) {}
+
+fn main() {
+    require_send(process(&1));
+}

